src/vstest.console/CommandLine/TestCaseFilter.cs

```diff
@@ -0,0 +1,194 @@
+                    if (testPropertyValueArray != null)
+                    {
+                        var testPropertyValueList = new List<string>(testPropertyValueArray);
+                        traitDictionary.Add(testPropertyKey, testPropertyValueList);
+                        continue;
+                    }
+                }
+
+                //always return value as a list of string
+                if (testPropertyValue != null)
+                {
+                    var multiValue = new List<string> { testPropertyValue.ToString() };
+                    traitDictionary.Add(testPropertyKey, multiValue);
+                }
+            }
+
+            return traitDictionary;
+        }
+
+        /// <summary>
+        /// fetch the traits on this test method and populate a trait dictionary
+        /// </summary>
+        private Dictionary<string, List<string>> GetTraitsInTraitDictionary(Dictionary<string, List<string>> traitDictionary, TraitCollection traits)
+        {
+            foreach (var trait in traits)
+            {
+                var newTraitValueList = new List<string> { trait.Value };
+                List<string> currentTraitValue;
+                if (!traitDictionary.TryGetValue(trait.Name, out currentTraitValue))
+                {
+                    // if the current trait's key is not already present, add the current trait key-value pair
+                    traitDictionary.Add(trait.Name, newTraitValueList);
+                }
+                else
+                {
+                    if (null == currentTraitValue)
+                    {
+                        // if the current trait's value is null, replace the previous value with the current value
+                        traitDictionary[trait.Name] = newTraitValueList;
+                    }
+                    else
+                    {
+                        // if the current trait's key is already present and is not null, append current value to the previous value list
+                        List<string> traitValueList = currentTraitValue;
+                        traitValueList.Add(trait.Value);
+                    }
+                }
+            }
+
+            return traitDictionary;
+        }
+
+        /// <summary>
+        /// Provides value for property name 'propertyName' as used in filter.
+        /// </summary>
+        private static string[] PropertyValueProvider(string propertyName, Dictionary<string, List<string>> traitDictionary)
+        {
+            List<string> propertyValueList;
+            traitDictionary.TryGetValue(propertyName, out propertyValueList);
+            if (propertyValueList != null)
+            {
+                var propertyValueArray = propertyValueList.ToArray();
+
+                return propertyValueArray;
+            }
+            return null;
+        }
+
+        private static TestCaseFilterExpression _filterExpression;
+
+        private const string TestCategory = "TestCategory";
```

Comment:
Nit: Make sure this class follows StyleCop standards.

Reply:
Done